mlflow/tracking/client.py

```diff
@@ -433,6 +436,222 @@ def get_trace_info(self, trace_id: str) -> TraceInfo:
+
+        if root_span_id is None:
+            raise MlflowException(f"Trace with ID {trace_id} not found.")
+
+        self.end_span(trace_id, root_span_id, inputs, outputs, attributes, status)
+
+    def start_span(
+        self,
+        name: str,
+        trace_id: str,
+        parent_span_id: str,
+        span_type: Optional[str] = None,
+        inputs: Optional[Dict[str, Any]] = None,
+        attributes: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Create a new span and start it without attaching it to the global trace context.
+
+        This is an imperative API to manually create a new span under a specific trace id and parent
+        span, unlike the higher-level APIs like `@mlflow.trace()` and `with mlflow.start_span()`,
+        which automatically manage the span lifecycle and parent-child relationship.
+
+        This API is useful for the case where the automatic context management is not sufficient,
+        such as callback-based instrumentation where span start and end are not in the same call
+        stack, or multi-threaded applications where the context is not propagated automatically.
+
+        To conclude the span for logging, call `span.end()`.
+
+        Args:
+            name: The name of the span.
+            trace_id: The ID of the trace to attach the span to.
+            span_type: The type of the span. Can be either a string or a SpanType enum value.
+            parent_span_id: The ID of the parent span. The parent span can be a span created by
+                both fluent APIs like `with mlflow.start_span()`, and imperative APIs like this.
+            inputs: Inputs to set on the span.
+            attributes: A dictionary of attributes to set on the span.
+
+        Returns:
+            An MLflowSpanWrapper object representing the span.
+
+        Example:
+
+        .. code-block:: python
+
+                from mlflow.tracking import MlflowClient
+
+                client = MlflowClient()
+
+                span = client.start_trace("my_trace")
+
+                x = 2
+
+                # Create a child span
+                child_span = client.start_span(
+                    "child_span",
+                    trace_id=span.trace_id,
+                    parent_span_id=span.id,
+                    inputs={"x": 2},
+                )
+
+                y = x**2
+
+                client.end_span(
+                    trace_id=child_span.trace_id,
+                    span_id=child_span.span_id,
+                    attributes={"factor": 2},
+                    outputs={"y": y},
+                )
+
+                client.end_trace(trace_id)
+        """
+        if not parent_span_id:
+            _logger.warning(
+                "start_span() must be called with an explicit parent_span_id."
+                "If you haven't start any span yet, use MLflowClient().start_trace() "
+                "to start a new trace and root span."
+            )
```

Comment:
This warning says `must` but it continues the rest, it's a little bit confusing to me

Reply:
Ah sorry I meant to return here. Indeed, I think it's fine to raise an exception here. This is not an internal error but the invalid use of tracing APIs.